packages/hash/api/src/graphql/resolvers/knowledge/link/link.ts

```diff
@@ -0,0 +1,114 @@
+import { ApolloError } from "apollo-server-errors";
+import { EntityModel, LinkModel, LinkTypeModel } from "../../../../model";
+import {
+  MutationCreateKnowledgeLinkArgs,
+  MutationDeleteKnowledgeLinkArgs,
+  QueryKnowledgeLinksArgs,
+  ResolverFn,
+} from "../../../apiTypes.gen";
+import { LoggedInGraphQLContext } from "../../../context";
+import {
+  mapLinkModelToGQL,
+  UnresolvedKnowledgeLinkGQL,
+} from "../model-mapping";
+
+export const createKnowledgeLink: ResolverFn<
+  Promise<UnresolvedKnowledgeLinkGQL>,
+  {},
+  LoggedInGraphQLContext,
+  MutationCreateKnowledgeLinkArgs
+> = async (_, { link }, { dataSources: { graphApi } }) => {
+  const { linkTypeId, ownedById, index, sourceEntityId, targetEntityId } = link;
+
+  const [linkTypeModel, sourceEntityModel, targetEntityModel] =
+    await Promise.all([
+      LinkTypeModel.get(graphApi, {
+        linkTypeId,
+      }),
+      EntityModel.getLatest(graphApi, {
+        entityId: sourceEntityId,
+      }),
+      EntityModel.getLatest(graphApi, {
+        entityId: targetEntityId,
+      }),
+    ]);
+
+  const linkModel = await LinkModel.create(graphApi, {
+    linkTypeModel,
+    ownedById,
+    index: index ?? undefined,
+    sourceEntityModel,
+    targetEntityModel,
+  });
+
+  return mapLinkModelToGQL(linkModel);
+};
+
+export const knowledgeLinks: ResolverFn<
+  Promise<UnresolvedKnowledgeLinkGQL[]>,
+  {},
+  LoggedInGraphQLContext,
+  QueryKnowledgeLinksArgs
+> = async (
+  _,
+  { sourceEntityId, linkTypeId },
+  { dataSources: { graphApi } },
+) => {
+  const linkModels = await LinkModel.getByQuery(graphApi, {
+    all: [
+      {
+        eq: [{ path: ["source", "id"] }, { literal: sourceEntityId }],
+      },
+      {
+        eq: [
+          { path: ["type", "versionedUri"] },
+          {
+            literal: linkTypeId,
+          },
+        ],
+      },
+    ],
+  });
+
+  return linkModels.map(mapLinkModelToGQL);
+};
+
+export const deleteKnowledgeLink: ResolverFn<
+  Promise<boolean>,
+  {},
+  LoggedInGraphQLContext,
+  MutationDeleteKnowledgeLinkArgs
+> = async (_, { link }, { dataSources: { graphApi }, user }) => {
+  const { linkTypeId, sourceEntityId, targetEntityId } = link;
+
+  const linkModels = await LinkModel.getByQuery(graphApi, {
+    all: [
+      { eq: [{ path: ["source", "id"] }, { literal: sourceEntityId }] },
+      { eq: [{ path: ["target", "id"] }, { literal: targetEntityId }] },
+      {
+        eq: [
+          { path: ["type", "versionedUri"] },
+          {
+            literal: linkTypeId,
+          },
+        ],
+      },
+    ],
+  });
+
+  if (!linkModels[0]) {
+    throw new ApolloError(
+      `Link with source enitty ID = '${sourceEntityId}', target entity ID = '${targetEntityId}' and link type ID = '${linkTypeId}' not found.`,
+      "NOT_FOUND",
+    );
+  } else if (linkModels.length > 1) {
+    throw new ApolloError(
+      `Could not identify one single link with query.`,
+      "BAD_REQUEST",
+    );
+  }
```

Comment:
Yes, i think it would make sense to encapsulate this in something like a `LinkModel.get` when we need `updateLink` (we might in `update-page-contents`)

Reply:
Can we do this now, as this is when the functionality is introduced? Otherwise if someone else is implementing `updateLink`, they may duplicate this implementation.

Reply:
Addressed in bcc8d0521fae3cac0b62dbdba3079fdf8d9363e1